cpp/src/join/cross_join.cu

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/detail/repeat.hpp>
+#include <cudf/detail/reshape.hpp>
+#include <cudf/filling.hpp>
+#include <cudf/join.hpp>
+#include <cudf/reshape.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_view.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/error.hpp>
+
+namespace cudf {
+namespace detail {
+/**
+ * @brief  Performs a cross join on two tables (left, right)
+ *
+ * The cross join returns the cartesian product of rows from each table.
+ *
+ * The approach is to repeat the left table by the number of rows in the right table
+ * and tile the right table by the number of rows in the left table.
+ *
+ * @throws cudf::logic_error if number of columns in either `left` or `right` table is 0
+ *
+ * @param[in] left                  The left table
+ * @param[in] right                 The right table
+ * @param[in] mr                    Device memory resource to use for device memory allocation
+ * @param[in] stream                Cuda stream
+ *
+ * @returns                         Result of cross joining `left` and `right` tables
+ */
+std::unique_ptr<cudf::table> cross_join(
+  cudf::table_view const& left,
+  cudf::table_view const& right,
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource(),
+  cudaStream_t stream                 = 0)
+{
+  CUDF_EXPECTS(0 != left.num_columns(), "Left table is empty");
+  CUDF_EXPECTS(0 != right.num_columns(), "Right table is empty");
+
+  // If left or right table has no rows, return an empty table with all columns
+  if ((0 == left.num_rows()) || (0 == right.num_rows())) {
+    auto empty_left_columns  = empty_like(left)->release();
+    auto empty_right_columns = empty_like(right)->release();
+    std::move(empty_right_columns.begin(),
```

Comment:
Wow. TIL `std::move` is also an algorithm: https://en.cppreference.com/w/cpp/algorithm/move

Reply:
I would have done this with a much uglier solution:
```
empty_left_columns.insert(empty_left_columns.end(), std::make_move_iterator(empty_right_columns.begin(), std::make_move_iterator(empty_right_columns.end());
```